derive_state_machine_future/src/ast.rs

```diff
@@ -154,6 +156,179 @@ where
+            &syn::Ty::Path(ref qself, ref p) => {
+                if let &Some(ref qself) = qself {
+                    qself.ty.collect_idents(idents);
+                }
+
+                p.collect_idents(idents);
+            }
+            &syn::Ty::Slice(ref ty) | &syn::Ty::Paren(ref ty) => ty.collect_idents(idents),
+            &syn::Ty::Ptr(ref ty) =>ty.ty.collect_idents(idents),
+            &syn::Ty::Rptr(ref lifetime, ref ty) => {
+                if let &Some(ref lifetime) = lifetime {
+                    idents.insert(lifetime.ident.clone());
+                }
+
+                ty.ty.collect_idents(idents)},
+            &syn::Ty::Tup(ref tys) => tys.iter().for_each(|v| v.collect_idents(idents)),
+            &syn::Ty::BareFn(ref bfn) => bfn.collect_idents(idents),
+            &syn::Ty::Array(ref ty, ref cexpr) => {
+                ty.collect_idents(idents);
+                cexpr.collect_idents(idents);
+            }
+            _ => {}
+        }
+    }
+}
+
+impl CollectIdents for syn::ConstExpr {
+    fn collect_idents(&self, idents: &mut HashSet<syn::Ident>) {
+        match self {
+            &syn::ConstExpr::Call(ref f, ref args) => {
+                f.collect_idents(idents);
+                args.iter().for_each(|v| v.collect_idents(idents));
+            }
+            &syn::ConstExpr::Binary(_, ref c0, ref c1) | &syn::ConstExpr::Index(ref c0, ref c1) => {
+                c0.collect_idents(idents);
+                c1.collect_idents(idents);
+            }
+            &syn::ConstExpr::Unary(_, ref c) | &syn::ConstExpr::Paren(ref c) => {
+                c.collect_idents(idents)
+            }
+            &syn::ConstExpr::Cast(ref c, ref ty) => {
+                ty.collect_idents(idents);
+                c.collect_idents(idents);
+            }
+            &syn::ConstExpr::Path(ref p) => p.collect_idents(idents),
+            _ => {}
+        }
+    }
+}
+
+impl CollectIdents for syn::BareFnTy {
+    fn collect_idents(&self, idents: &mut HashSet<syn::Ident>) {
+        self.inputs.iter().for_each(|v| v.ty.collect_idents(idents));
+
+        match self.output {
+            syn::FunctionRetTy::Ty(ref ty) => ty.collect_idents(idents),
+            _ => {}
+        }
+    }
+}
+
+impl CollectIdents for syn::Path {
+    fn collect_idents(&self, idents: &mut HashSet<syn::Ident>) {
+        if self.segments.len() == 1 {
+            // if the path contains only one element, it could be a generic type parameter
+            let last = self.segments.get(0).unwrap();
+            idents.insert(last.ident.clone());
+            last.parameters.collect_idents(idents);
+        } else {
+            // only collect the idents of the path parameters (aka type/lifetime parameters),
+            // because we don't want to collect the "std::mem" in "std::mem::some_function"
+            self.segments
+                .iter()
+                .for_each(|s| s.parameters.collect_idents(idents));
```

Comment:
Yeah, you are right, the else branch is doing almost the same. 
The following is the only difference:
```
let last = self.segments.get(0).unwrap();
idents.insert(last.ident.clone());
```

Yeah only the first element can be a type parameter, if the number of segments is one. You are right, I also need to check for `self.global`. If we have more than one segment, we have a type with a Path and that can not be a type parameter. However, we still need to check the parameters, because that could be one of our type/lifetime parameters.